ingester/user_state.go

```diff
@@ -183,11 +184,23 @@ func (u *userState) unlockedGet(metric model.Metric, maxSeries int) (model.Finge
 	// all proceed to add a new series. This is likely not worth addressing,
 	// as this should happen rarely (all samples from one push are added
 	// serially), and the overshoot in allowed series would be minimal.
-	if u.fpToSeries.length() >= maxSeries {
+	if u.fpToSeries.length() >= cfg.MaxSeriesPerUser {
 		u.fpLocker.Unlock(fp)
 		return fp, nil, util.ErrUserSeriesLimitExceeded
 	}
 
+	metricName, err := util.ExtractMetricNameFromMetric(metric)
+	if err != nil {
+		u.fpLocker.Unlock(fp)
+		return fp, nil, err
+	}
+
+	// The same race can happen here as with series-per-user above.
+	if u.index.numSeriesInMetric(metricName) >= cfg.MaxSeriesPerMetric {
```

Comment:
I pushed another commit which does this. See how you like it. It requires one more mutex and one more explicit call to `util.ExtractMetricName`. I'm not worried about speed, since it's only in the sad path, but wondering if it's nicer this way or not.